File: Games/Dennis/src/de/hsh/dennis/model/NpcLogic/NpcIO.java
package de.hsh.dennis.model.NpcLogic;

import com.google.gson.Gson;
import common.util.Logger;
import de.hsh.dennis.model.NpcLogic.actors.Bot;
import de.hsh.dennis.model.NpcLogic.actors.Hacker;
import de.hsh.dennis.model.NpcLogic.actors.Npc;
import de.hsh.dennis.model.NpcLogic.actors.Package;

import java.io.*;

public class NpcIO {

    /**
     * Gets the execution location of the main program.
     *
     * @return a String representation.
     *
     * @author Alexander Komischke
     */
    public static String getExecutionLocation() {
        return NpcIO.class.getProtectionDomain().getCodeSource().getLocation().getPath();
    }

    public Npc[] loadLevel(SkinConfig.Level.Difficulty dif) {

        //JsonNpc[] tempArray = new JsonNpc[]{new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE)};
        //gsonWrite(tempArray, "temp");

        String fileName;
        switch (dif) {
            case MEDIUM:
                fileName = "medium";
                break;
            case HARD:
                fileName = "hard";
                break;
            case NIGHTMARE:
                fileName = "nightmare";
                break;
            case CUSTOM:
                fileName = "temp";
                break;
            default:
                fileName = "easy";
                break;
        }

        return gsonRead(fileName);
    }

    private Npc[] gsonRead(String fileName) {
        BufferedReader bufferedReader = null;
        try {
            bufferedReader = new BufferedReader(new FileReader(getExecutionLocation() + "de/hsh/dennis/resources/levelFiles/" + fileName + ".json"));
            return readNPC(bufferedReader);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } finally {
            if (bufferedReader != null) {
                try {
                    bufferedReader.close();
                } catch (IOException ioException) {
                    ioException.printStackTrace();
                }
            }
        }
        return new Npc[]{};
    }

    private Npc[] readNPC(BufferedReader bufferedReader) {
        Gson gson = new Gson();
        JsonNpc[] parsingArray = gson.fromJson(bufferedReader, JsonNpc[].class);
        Npc[] returnArray = new Npc[parsingArray.length];
        for (int i = 0; i < parsingArray.length; i++) {

            if (parsingArray[i].getNpcType() != null) {
                switch (parsingArray[i].getNpcType()) {
                    case PACKAGE:
                        returnArray[i] = new Package(parsingArray[i].getSpawnType(), parsingArray[i].getSpawnTime(), parsingArray[i].getSpeed());
                        break;
                    case BOT:
                        returnArray[i] = new Bot(parsingArray[i].getSpawnType(), parsingArray[i].getSpawnTime(), parsingArray[i].getSpeed());
                        break;

                    case HACKER:
                        returnArray[i] = new Hacker(parsingArray[i].getSpawnType(), parsingArray[i].getSpawnTime(), parsingArray[i].getSpeed());
                        break;
                    default:
                        Logger.log(this.getClass().getName() + "failed to create matching NpcLogic");
                }
            } else returnArray[i] = null;
        }

        return returnArray;
    }
}

File: Games/Dennis/src/de/hsh/dennis/model/NpcLogic/NpcIO.kt
package de.hsh.dennis.model.NpcLogic

import com.google.gson.Gson
import common.util.Logger.log
import de.hsh.dennis.model.NpcLogic.NPCEnums.NpcType
import de.hsh.dennis.model.NpcLogic.SkinConfig.Level.Difficulty
import de.hsh.dennis.model.NpcLogic.actors.Bot
import de.hsh.dennis.model.NpcLogic.actors.Hacker
import de.hsh.dennis.model.NpcLogic.actors.Npc
import de.hsh.dennis.model.NpcLogic.actors.Package
import java.io.BufferedReader
import java.io.FileNotFoundException
import java.io.FileReader
import java.io.IOException

class NpcIO {
    fun loadLevel(dif: Difficulty?): Array<Npc?> {

        //JsonNpc[] tempArray = new JsonNpc[]{new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE), new JsonNpc(1.0d, 1.0d, NPCEnums.Spawn.RIGHT, NPCEnums.NpcType.PACKAGE)};
        //gsonWrite(tempArray, "temp");
        val fileName: String
        fileName = when (dif) {
            Difficulty.MEDIUM    -> "medium"
            Difficulty.HARD      -> "hard"
            Difficulty.NIGHTMARE -> "nightmare"
            Difficulty.CUSTOM    -> "temp"
            else                 -> "easy"
        }
        return gsonRead(fileName)
    }

    private fun gsonRead(fileName: String): Array<Npc?> {
        var bufferedReader: BufferedReader? = null
        try {
            bufferedReader = BufferedReader(FileReader(executionLocation + "de/hsh/dennis/resources/levelFiles/" + fileName + ".json"))
            return readNPC(bufferedReader)
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } finally {
            if (bufferedReader != null) {
                try {
                    bufferedReader.close()
                } catch (ioException: IOException) {
                    ioException.printStackTrace()
                }
            }
        }
        return arrayOf()
    }

    private fun readNPC(bufferedReader: BufferedReader): Array<Npc?> {
        val gson = Gson()
        val parsingArray = gson.fromJson(bufferedReader,
                                         Array<JsonNpc>::class.java)
        val returnArray = arrayOfNulls<Npc>(parsingArray.size)
        for (i in parsingArray.indices) {
            when (parsingArray[i].npcType) {
                NpcType.PACKAGE -> returnArray[i] = Package(parsingArray[i].spawnType,
                                                            parsingArray[i].spawnTime,
                                                            parsingArray[i].speed)
                NpcType.BOT     -> returnArray[i] = Bot(parsingArray[i]
                                                                .spawnType,
                                                        parsingArray[i]
                                                                .spawnTime,
                                                        parsingArray[i]
                                                                .speed)
                NpcType.HACKER  -> returnArray[i] = Hacker(parsingArray[i].spawnType,
                                                           parsingArray[i].spawnTime,
                                                           parsingArray[i].speed)
            }
        }
        return returnArray
    }

    companion object {
        /**
         * Gets the execution location of the main program.
         *
         * @return a String representation.
         *
         * @author Alexander Komischke
         */
        val executionLocation: String
            get() = NpcIO::class.java.protectionDomain.codeSource.location.path
    }
}
